python/descriptor_pool.c

```diff
@@ -146,6 +190,45 @@ static PyObject* PyUpb_DescriptorPool_AddSerializedFile(
     goto done;
   }
 
+  upb_strview name = google_protobuf_FileDescriptorProto_name(proto);
+  const upb_filedef* file =
+      upb_symtab_lookupfile2(self->symtab, name.data, name.size);
+
+  if (file) {
+    // If the existing file is equal to the new file, then silently ignore the
+    // duplicate add.
+    google_protobuf_FileDescriptorProto* existing =
+        upb_FileDef_ToProto(file, arena);
+    if (!existing) {
+      PyErr_SetNone(PyExc_MemoryError);
+      goto done;
+    }
+    const upb_msgdef* m = PyUpb_DescriptorPool_GetFileProtoDef();
+    if (PyUpb_Message_IsEqual(proto, existing, m)) {
+      Py_INCREF(Py_None);
+      result = Py_None;
+      goto done;
+    }
+  }
+
+  if (self->db) {
+    // Load dependent files if necessary.
+    size_t n;
+    const upb_strview* deps =
+        google_protobuf_FileDescriptorProto_dependency(proto, &n);
+    for (size_t i = 0; i < n; i++) {
+      const upb_filedef* dep =
+          upb_symtab_lookupfile2(self->symtab, deps[i].data, deps[i].size);
+      if (!dep) {
+        PyObject* filename =
+            PyUnicode_FromStringAndSize(deps[i].data, deps[i].size);
+        if (!filename) goto done;
+        if (!PyUpb_DescriptorPool_TryLoadFilename(self, filename)) goto done;
```

Comment:
the failure case here is missing a decref on filename

Reply:
Done.